packages/reference/src/common/EntityStore.ts

```diff
@@ -108,101 +191,157 @@ function useEntitiesStore(props: { sdk: BaseExtensionSDK }) {
+      try {
+        const entry = await cmaClient.entry.get({ spaceId, environmentId, entryId });
+        dispatch({ type: 'set_entry', id: entryId, entry });
+        return entry;
+      } catch (error) {
+        dispatch({ type: 'set_entry_failed', id: entryId });
+        return;
+      }
     },
-    [props.sdk.space]
+    [cmaClient, spaceId, environmentId]
   );
 
-  const loadAsset = React.useCallback(
-    (id: string) => {
-      return props.sdk.space
-        .getAsset(id)
-        .then((asset) => {
-          dispatch({ type: 'set_asset', id, asset });
-          return asset;
-        })
-        .catch(() => {
-          dispatch({ type: 'set_asset_failed', id });
-        });
+  const getEntry = React.useCallback(
+    async (entryId: string) => {
+      const cachedEntry = state.entries[entryId];
+
+      if (isNotNil(cachedEntry)) {
+        return cachedEntry;
+      }
+
+      return loadEntry(entryId);
     },
-    [props.sdk.space]
+    [loadEntry, state.entries]
   );
 
-  const getOrLoadAsset = React.useCallback(
-    (id: string) => {
-      if (state.assets[id] && state.assets[id] !== 'failed') {
-        return Promise.resolve(state.assets[id]);
+  const loadAsset = React.useCallback(
+    async (assetId: string) => {
+      try {
+        const asset = await cmaClient.asset.get({ spaceId, environmentId, assetId });
+        dispatch({ type: 'set_asset', id: assetId, asset });
+        return asset;
+      } catch (error) {
+        dispatch({ type: 'set_asset_failed', id: assetId });
+        return;
       }
-      return loadAsset(id);
     },
-    [state.assets, loadAsset]
+    [cmaClient, spaceId, environmentId]
   );
 
-  const getOrLoadEntry = React.useCallback(
-    (id: string) => {
-      if (state.entries[id] && state.entries[id] !== 'failed') {
-        return Promise.resolve(state.entries[id]);
+  const getAsset = React.useCallback(
+    async (assetId: string) => {
+      const cachedAsset = state.assets[assetId];
+
+      if (isNotNil(cachedAsset)) {
+        return cachedAsset;
       }
-      return loadEntry(id);
+
+      return loadAsset(assetId);
+    },
+    [loadAsset, state.assets]
+  );
+
+  const loadContentfulEntry = React.useCallback(
+    async (urn: string): Promise<ResourceInfo<Entry>> => {
+      const resourceId = urn.split(':', 6)[5];
+      const [, spaceId, , entryId] = resourceId.split('/');
+      const environmentId = 'master';
+      const [space, entry] = await Promise.all([
+        await cmaClient.space.get({ spaceId }),
+        await cmaClient.entry.get({ spaceId, environmentId, entryId }),
+      ]);
+      const contentTypeId = entry.sys.contentType.sys.id;
+      const [contentType, locales] = await Promise.all([
+        await cmaClient.contentType.get({
+          contentTypeId,
+          spaceId,
+          environmentId,
+        }),
+        await cmaClient.locale.getMany({ spaceId, environmentId, query: { limit: 100 } }),
+      ]);
```

Comment:
While waiting for real optimizations:
```suggestion
      const entryPromise = cmaClient.entry.get({ spaceId, environmentId, entryId })
      const contentTypePromise = entryPromise.then((entry) => {
          const contentTypeId = entry.sys.contentType.sys.id;
          return cmaClient.contentType.get({
          contentTypeId,
          spaceId,
          environmentId,
        }),
      })
      const [space, entry, contentType, locales] = await Promise.all([
        await cmaClient.space.get({ spaceId }),
        entryPromise,
        contentTypePromise,
        await cmaClient.locale.getMany({ spaceId, environmentId, query: { limit: 100 } })
      ]);
```